### PR TITLE
Implement sync function of legacy chain handler - Closes  #7547 & #7486

### LINKED ISSUE #7486: Implement syncBlocks function of LegacyChainHandler
### Description

Implement `syncBlocks` function of `LegacyChainHandler`
1. Get a list of connectedPeers from the network
2.  Find peers with legacy info corresponding to `snapshotBlockID` (if it contains `legacy` property with the `snapshotBlockID` as requested then it means it has the legacy blocks history)
3. If no peer is found then throw `PeerNotFound` error
4. Randomly select a peer from the filtered peers list (something like `Math.trunc(Math.random() * filteredPeers.length - 1)`)
5. Request the peer with `getLegacyBlocksFromId` RPC and receive response with request param `{ id: Buffer }`
4.1. if response is invalid or `validateLegacyBlock(blockBytes, nextBlock)` fails then ban the peer with 100 penalty by calling `network.applyPenalty(peerID, banScore)`
6.  Save the blocks and repeat step 1 to 4 until we reach startHeight defined in the config.
7.  Once we reach the startHeight of a bracket we save the bracketInfo and return `setLegacyChainBracketInfo(bracketInfo)`

### Acceptance Criteria

- Should have all the cases covered with the unit tests 

### Addtional Information

- Depends on #7504
- Depends on https://github.com/LiskHQ/lisk-sdk/issues/7498
- Define `PeerNotFound` error with argument as message 
- Relates to https://github.com/LiskHQ/lisk-sdk/issues/7547

### LINKED ISSUE #7547: Implement sync method on LegacyChainHandler
### Description

- Implement `sync` method
  1. Iterate through `brackets` and call `bracketInfo = getLegacyChainBracketInfo(bracket.snapshotBlockID)`
  2. If `bracket.startHeight === bracketInfo.lastHeight` then skip to next bracket and call step 1
  3. If `bracket.startHeight !== bracketInfo.lastHeight` then call `lastBlock = getBlockByHeight(bracketInfo.lastHeight)`
  4. Call `syncBlocks(bracket, lastBlock)`, if `syncBlocks` function throws `PeerNotFound` error then wait for 2 minutes (to let enough shuffling to happen to find peer with legacyBlocks) and call this function again
  5. If syncBlocks function returns then continue with next bracket and call step 1 until all the brackets are synced and then call `applyNodeInfo({ legacy: brackets.map(b => b.snapshotBlockID )})` with all the bracket's snapshotBlockID

### Acceptance Criteria

- Should have unit test for sync method and cover all the cases

### Additional Information

- Depends on #7545
- Relates to https://github.com/LiskHQ/lisk-sdk/issues/7486
- https://github.com/LiskHQ/lisk-sdk/issues/7674

### PR DESCRIPTION
**Pull request by sitetester**
### What was the problem?

This PR resolves #7547 and resolves #7486

### How was it solved?

`syncBlocks()` function is commented with flow of syncing legacy blocks

### CODE REVIEW

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 1]
```diff
@@ -12,41 +12,170 @@
...
+
+		for (const bracket of this._legacyConfig.brackets) {
+			const bracketInfo = await this._storage.getLegacyChainBracketInfo(bracket.snapshotBlockID);
+			if (bracket.startHeight === bracketInfo.lastHeight) {
+				continue;
+			}
+			const lastBlock = await this._storage.getBlockByHeight(bracketInfo.lastHeight);
+			const { id } = decodeBlock(lastBlock).block.header;
+			await this._trySyncBlocks(bracket, Buffer.from(id).toString('hex'));
+			await this._storage.saveBracketInfo(bracketInfo);
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:64]: here we don't set updated bracketInfo so we need to save the updated one after the sync

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 2]
```diff
@@ -12,41 +12,170 @@
...
+		});
+	}
+
+	private async _trySyncBlocks(bracket: LegacyBlockBracket, lastBlock: string) {
+		try {
+			await this.syncBlocks(bracket, lastBlock);
+		} catch (err) {
+			if (err instanceof PeerNotFoundWithLegacyInfo) {
+				// eslint-disable-next-line @typescript-eslint/no-misused-promises
+				setTimeout(async () => {
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:78]: create a private variable in this class to capture setTimeout return value and use it to clearTimeout after we finish sync.

**Review by ishantiw — CHANGES_REQUESTED** [review]
Overall approach looks good to me, few comments

File: framework/test/unit/engine/legacy/legacy_chain_handler.spec.ts [hunk 3]
```diff
@@ -48,7 +122,17 @@ describe('Legacy Chain Handler', () => {
 		});
 	});
 
-	describe('syncBlocks', () => {
-		it.todo('test behaviors');
+	describe('sync', () => {
+		it('test behaviors', async () => {
```
Comment (mosmartin) [inline framework/test/unit/engine/legacy/legacy_chain_handler.spec.ts:126]: I've noticed you've not used the convention `it(should...)` as well as you've not defined the behavior that is being tested.

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 4]
```diff
@@ -12,56 +12,193 @@
...
-	db: Database | InMemoryDatabase;
+	db: Database;
 }
 
 export class LegacyChainHandler {
+	private readonly _network: Network;
+	private _storage!: Storage;
 	private readonly _legacyConfig: LegacyConfig;
-	private _db!: Database | InMemoryDatabase;
+	private _timeout!: NodeJS.Timeout;
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:41]: maybe give more meaning name here

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 5]
```diff
@@ -12,56 +12,193 @@
...
-				Buffer.from(bracket.snapshotBlockID, 'hex'),
-			);
-
-			if (!bracketInfo) {
-				await storage.setLegacyChainBracketInfo(Buffer.from(bracket.snapshotBlockID), {
+			try {
+				await this._storage.getLegacyChainBracketInfo(Buffer.from(bracket.snapshotBlockID, 'hex'));
+			} catch (err) {
+				// Save config brackets in advance, these will be used in next step (`sync`)
+				await this._storage.setLegacyChainBracketInfo(Buffer.from(bracket.snapshotBlockID), {
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:56]: If there is any other error apart from key not found we should handle it 
```ts
if (!(error instanceof NotFoundError)) {
				throw error;
			}
```

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 6]
```diff
@@ -12,56 +12,193 @@
...
+				continue;
+			}
+			const lastBlock = decodeBlock(
+				await this._storage.getBlockByHeight(bracketInfo.lastBlockHeight),
+			).block;
+
+			await this._trySyncBlocks(bracket, lastBlock);
+			await this._storage.setLegacyChainBracketInfo(
+				Buffer.from(bracket.snapshotBlockID, 'hex'),
+				bracketInfo,
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:88]: you need to update the bracket info with the last block after syncing which will be the block at the startHeight. Its better to set bracket everytime we finish one call with the peer, so once we process 100 blocks received, we should set bracket and then call another peer for next 100 blocks.

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 7]
```diff
@@ -12,56 +12,193 @@
...
+					block.header.id as Buffer,
+					block.header.height,
+					encodeBlock(block),
+				);
+				lastBlock = block;
+			}
+		}
+
+		if (lastBlock && lastBlock.header.height > bracket.startHeight) {
+			await this.syncBlocks(bracket, lastBlock);
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:185]: we can add setbracket with updated bracketInfo before calling syncBlocks

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 8]
```diff
@@ -12,56 +12,193 @@
...
+		}
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:186]: before we exit the sync , we update bracketInfo

File: framework/test/unit/engine/legacy/legacy_chain_handler.spec.ts [hunk 9]
```diff
@@ -14,32 +14,144 @@
...
+			},
+			{
+				peerId: 'peerId-2',
+				options: {
+					legacy: [randomSnapshotBlockID, expectedSnapshotBlockID],
 				},
+			},
+		];
+		// CAUTION: This is copy of blockFixtures[0](19583715), ONLY height has changed
+		legacyBlock19583716 = {
```
Comment (ishantiw) [inline framework/test/unit/engine/legacy/legacy_chain_handler.spec.ts:73]: best would be to move this to fixture file

**Review by ishantiw — CHANGES_REQUESTED** [review]

**Review by mosmartin — COMMENTED** [review]
Besides the open comments, LGTM 🚀

**Review by mosmartin — APPROVED** [review]
Having another look.

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 10]
```diff
@@ -12,56 +12,202 @@
...
-		// TODO: implement the logic, general flow is as below
-		// 1. check config if sync legacy flag is true
-		// 2. Checks to what height blocks are present for a snapshotBlockID backwards
-		// 3. checks for all the peers node info if any peer has legacy blocks for the snapshot block ID, if not wait for X second
-		// 4. sync the blocks
-		// 5. Update node info
+	public async sync() {
+		if (!this._legacyConfig.sync) {
+			return;
+		}
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:71]: this is already checked in engine so we don't need to check here again

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 11]
```diff
@@ -12,56 +12,202 @@
...
+	 * Check if we have `sync` property `true` in configuration
+	 * getConnectedPeers from network
+	 * Filter peers having their `legacy` buffer array property contains `snapshotBlockID`
+	 * If there is no peer, throw error, this error will be used in outside function to retry calling `syncBlocks` after x amount of time
+	 * Get a random peer from list of filtered peers with legacy info
+	 * Make a request to that random peer by calling its `getLegacyBlocksFromId` method with `data` property set to `legacyBlock.header.id`
+	 * Try to decode response data buffer into Block, apply penalty in case of error
+	 * Start saving parsed blocks one by one
+	 * repeat, if last parsed block height is still higher than bracket.startHeight
+	 * finally, update node info
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:127]: I think the description needs to be updated according to the latest code

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 12]
```diff
@@ -12,56 +12,202 @@
...
+			// this part is needed to make sure `data` returns ONLY `{ blocks: Buffer[] }` & not any extra field(s)
+			const { blocks } = codec.decode<{ blocks: Buffer[] }>(
+				getBlocksFromIdResponseSchema,
+				data as Buffer,
+			);
+			this._applyValidation(blocks);
+			legacyBlocks = blocks.map(block => decodeBlock(block).block);
+		} catch (err) {
+			this._network.applyPenaltyOnPeer({ peerId, penalty: 100 });
+			return;
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:168]: I think here it should throw error instead of return otherwise , it will continue to the syncing next bracket without finishing this bracket

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 13]
```diff
@@ -12,56 +12,202 @@
...
+		const randomPeerIndex = Math.trunc(Math.random() * peersWithLegacyInfo.length - 1);
+		const { peerId } = peersWithLegacyInfo[randomPeerIndex];
+
+		const p2PRequestPacket: P2PRequestPacket = {
+			procedure: 'getLegacyBlocksFromId',
+			data: legacyBlock.header.id,
+		};
+
+		const response = await this._network.requestFromPeer({ ...p2PRequestPacket, peerId });
+		if (!(response.data as []).length) {
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:151]: I think this condition seems irrelevant, as we are checking response.data which is a buffer that is decoded at line 165

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 14]
```diff
@@ -12,56 +12,202 @@
...
+		const { data } = response;
+		let legacyBlocks: LegacyBlock[];
+		try {
+			// this part is needed to make sure `data` returns ONLY `{ blocks: Buffer[] }` & not any extra field(s)
+			const { blocks } = codec.decode<{ blocks: Buffer[] }>(
+				getBlocksFromIdResponseSchema,
+				data as Buffer,
+			);
+			this._applyValidation(blocks);
+			legacyBlocks = blocks.map(block => decodeBlock(block).block);
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:165]: i think here you  should check whether blocks.length >0 , else you should applyPenalty 100 and this._network.applyPenaltyOnPeer({ peerId, penalty: 100 and call syncBlocks instead of return

File: framework/src/engine/legacy/legacy_chain_handler.ts (hunk 12 above)
Comment (ishantiw) [reply framework/src/engine/legacy/legacy_chain_handler.ts:168]: so in general this code block should be like,
prefer this being called and applied penalty separately,

- check blocks.length > 0, apply penalty, log that penalty was applied to a peer and call sync() again
- if the above check passes, call _applyValidation, log that penalty was applied to a peer, if fails then apply penalty and call sync() again

**Review by ishantiw — CHANGES_REQUESTED** [review]

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 15]
```diff
@@ -15,6 +15,7 @@
 import { codec } from '@liskhq/lisk-codec';
 import { P2PRequestPacket } from '@liskhq/lisk-p2p/dist-node/types';
 import { Database, NotFoundError } from '@liskhq/lisk-db';
+import { INVALID_PEER_INFO_PENALTY } from '@liskhq/lisk-p2p/dist-node/constants';
```
Comment (ishantiw) [inline framework/src/engine/legacy/legacy_chain_handler.ts:18]: wrong import and maybe we should simply use 100 penalty or declare a constant within legacy to use

**Review by ishantiw — CHANGES_REQUESTED** [review]
Just one point otherwise lgtm

**Review by ishantiw — APPROVED** [review]
Great job @sitetester 👏

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 16]
```diff
@@ -12,56 +12,215 @@
...
+			await this.syncBlocks(bracket, legacyBlock);
+		};
+
+		try {
+			// this part is needed to make sure `data` returns ONLY `{ blocks: Buffer[] }` & not any extra field(s)
+			const { blocks } = codec.decode<{ blocks: Buffer[] }>(
+				getBlocksFromIdResponseSchema,
+				data as Buffer,
+			);
+			if (!(blocks.length > 0)) {
```
Comment (mosmartin) [inline framework/src/engine/legacy/legacy_chain_handler.ts:168]: For this ` if (!(blocks.length > 0)) ` piece of logic, would it be better to use the opposite operator (<=) `if (blocks.length <= 0)` instead? Not a show stopper by any means, just a suggestion.

File: framework/src/engine/legacy/legacy_chain_handler.ts [hunk 17]
```diff
@@ -12,56 +12,215 @@
...
+				await applyPenaltyAndRepeat();
+			}
+
+			this._applyValidation(blocks);
+
+			legacyBlocks = blocks.map(block => decodeBlock(block).block);
+			if (!(legacyBlocks.length > 0)) {
```
Comment (mosmartin) [inline framework/src/engine/legacy/legacy_chain_handler.ts:175]: For this `if (!(legacyBlocks.length > 0))` piece of logic, would it be better to use the opposite operator (<=) `if (legacyBlocks.length <= 0)` instead? Not a show stopper by any means, just a suggestion.
